### PR TITLE
Feature/epos4 support

### PR DESCRIPTION
**Pull request by jabbenseth**

### CODE REVIEW

File: epos_hardware/install_udev_rules [hunk 1]
```diff
@@ -4,11 +4,13 @@ echo ""
 echo "This script copies a udev rule for ftd2xx devices"
 echo ""
 
-sudo cp `rospack find epos_hardware`/90-ftd2xx.rules /etc/udev/rules.d
+sudo cp `rospack find epos_hardware`/rules/90-ftd2xx.rules /etc/udev/rules.d
+sudo cp `rospack find epos_hardware`/rules/99-epos4.rules /etc/udev/rules.d
+sudo cp `rospack find epos_hardware`/rules/99-ftdi.rules /etc/udev/rules.d
```
Comment (ipa-fez) [inline epos_hardware/install_udev_rules]: these are both `99`, do you want to differentiate the numbers to make the order explicit?

File: epos_hardware/src/tools/list_devices.cpp [hunk 2]
```diff
@@ -33,39 +33,39 @@ int main(int argc, char** argv){
       std::vector<std::string> protocol_stack_names;
       if(GetProtocolStackNameList(device_name, &protocol_stack_names, &error_code)) {
 	BOOST_FOREACH(const std::string& protocol_stack_name, protocol_stack_names) {
```
Comment (ipa-fez) [inline epos_hardware/src/tools/list_devices.cpp:35]: No `c++11` available?

File: epos_hardware/src/util/epos.cpp [hunk 3]
```diff
@@ -108,85 +155,114 @@ class ParameterSetLoader {
...
-  if(config_nh_.getParam("fault_reaction_option", fault_reaction_str)) {
-    if(fault_reaction_str == "signal_only") {
+#define SET_FAULT_REACTION_OPTION(val)                                                                                 \
+  do                                                                                                                   \
+  {                                                                                                                    \
+    unsigned int length = 2;                                                                                           \
+    unsigned int bytes_written;                                                                                        \
+    int16_t data = val;                                                                                                \
+    VCS(SetObject, 0x605E, 0x00, &data, length, &bytes_written);                                                       \
+  } while (true)
```
Comment (ipa-fez) [inline epos_hardware/src/util/epos.cpp:240]: Man, all these macros are pretty ugly...

File: epos_hardware/src/util/epos.cpp [hunk 4]
```diff
@@ -310,292 +393,336 @@ bool Epos::init() {
...
+  //       return false;
+  //     if (velocity_profile_window)
+  //     {
+  //       VCS(EnableVelocityWindow,
+  //           window,
+  //           (int)(1000 * time)  // s -> ms
+  //           );
+  //     }
+  //   }
+  // }
```
Comment (ipa-fez) [inline epos_hardware/src/util/epos.cpp]: delete?

File: epos_hardware/src/util/epos.cpp [hunk 5]
```diff
@@ -310,292 +393,336 @@ bool Epos::init() {
     VCS_FROM_SINGLE_PARAM_OPTIONAL(safety_nh, int, max_following_error, SetMaxFollowingError);
     VCS_FROM_SINGLE_PARAM_OPTIONAL(safety_nh, int, max_profile_velocity, SetMaxProfileVelocity);
     VCS_FROM_SINGLE_PARAM_OPTIONAL(safety_nh, int, max_acceleration, SetMaxAcceleration);
-    if(max_profile_velocity_set)
+    if (max_profile_velocity_set)
       max_profile_velocity_ = max_profile_velocity;
     else
       max_profile_velocity_ = -1;
   }
```
Comment (ipa-fez) [inline epos_hardware/src/util/epos.cpp:401]: I like all these extra scopes, why not break all these blocks out into functions?

File: epos_hardware/src/util/epos.cpp [hunk 6]
```diff
@@ -310,292 +393,336 @@ bool Epos::init() {
...
       return;
-    VCS_MoveToPosition(node_handle_->device_handle->ptr, node_handle_->node_id, (int)position_cmd_, true, true, &error_code);
+    long cmd = static_cast<long>(velocity_cmd_ * 60 / (2 * M_PI));
+    if (max_profile_velocity_ > 0) // Allow clamping of
+    {
+      if (cmd < -max_profile_velocity_ * 60 / (2 * M_PI))
+        cmd = -max_profile_velocity_ * 60 / (2 * M_PI);
+      if (cmd > max_profile_velocity_ * 60 / (2 * M_PI))
+        cmd = max_profile_velocity_ * 60 / (2 * M_PI);
+    }
```
Comment (ipa-fez) [inline epos_hardware/src/util/epos.cpp]: Duplicate code, also I think there's a `boost::clamp` or something...

**Review by ipa-fez — CHANGES_REQUESTED** [review]

**Comment by jabbenseth** [comment]
I will integrate some of your comments. however I will not start a major rework (scopes,macros, ...) Many changes are due to formatting

File: epos_hardware/src/util/epos.cpp [hunk 7]
```diff
@@ -59,44 +91,59 @@ Epos::Epos(const std::string& name,
 
   diagnostic_updater_.setHardwareID(serial_number_str);
   std::stringstream motor_diagnostic_name_ss;
-  motor_diagnostic_name_ss << name << ": " << "Motor";
+  motor_diagnostic_name_ss << name << ": "
+                           << "Motor";
```
Comment (benmaidel) [inline epos_hardware/src/util/epos.cpp:96]: Why did you put this to a new line?

File: epos_hardware/src/util/epos.cpp [hunk 8]
```diff
@@ -59,44 +91,59 @@ Epos::Epos(const std::string& name,
...
   diagnostic_updater_.add(motor_diagnostic_name_ss.str(), boost::bind(&Epos::buildMotorStatus, this, _1));
   std::stringstream motor_output_diagnostic_name_ss;
-  motor_output_diagnostic_name_ss << name << ": " << "Motor Output";
+  motor_output_diagnostic_name_ss << name << ": "
+                                  << "Motor Output";
```
Comment (benmaidel) [inline epos_hardware/src/util/epos.cpp:100]: Why did you put this to a new line?

File: epos_library/include/epos_library/Definitions.h [hunk 9]
```diff
@@ -59,15 +57,16 @@
     Initialisation_DllExport int  VCS_CloseAllDevices(unsigned int* pErrorCode);
 
 //Info
-    HelpFunctions_DllExport int  VCS_GetVersion(void* KeyHandle, unsigned short NodeId, unsigned short* pHardwareVersion, unsigned short* pSoftwareVersion, unsigned short* pApplicationNumber, unsigned short* pApplicationVersion, unsigned int* pErrorCode);
-    HelpFunctions_DllExport int  VCS_GetErrorInfo(unsigned int ErrorCodeValue, char* pErrorInfo, unsigned short MaxStrSize);
     HelpFunctions_DllExport int  VCS_GetDriverInfo(char* p_pszLibraryName, unsigned short p_usMaxLibraryNameStrSize,char* p_pszLibraryVersion, unsigned short p_usMaxLibraryVersionStrSize, unsigned int* p_pErrorCode);
+	HelpFunctions_DllExport int  VCS_GetVersion(void* KeyHandle, unsigned short NodeId, unsigned short* pHardwareVersion, unsigned short* pSoftwareVersion, unsigned short* pApplicationNumber, unsigned short* pApplicationVersion, unsigned int* pErrorCode);
```
Comment (benmaidel) [inline epos_library/include/epos_library/Definitions.h:61]: Tab vs. Spaces

File: epos_library/include/epos_library/Definitions.h [hunk 10]
```diff
@@ -59,15 +57,16 @@
...
+    HelpFunctions_DllExport int  VCS_GetErrorInfo(unsigned int ErrorCodeValue, char* pErrorInfo, unsigned short MaxStrSize);
 
 //Advanced Functions
     HelpFunctions_DllExport int  VCS_GetDeviceNameSelection(int StartOfSelection, char* pDeviceNameSel, unsigned short MaxStrSize, int* pEndOfSelection, unsigned int* pErrorCode);
     HelpFunctions_DllExport int  VCS_GetProtocolStackNameSelection(char* DeviceName, int StartOfSelection, char* pProtocolStackNameSel, unsigned short MaxStrSize, int* pEndOfSelection, unsigned int* pErrorCode);
     HelpFunctions_DllExport int  VCS_GetInterfaceNameSelection(char* DeviceName, char* ProtocolStackName, int StartOfSelection, char* pInterfaceNameSel, unsigned short MaxStrSize, int* pEndOfSelection, unsigned int* pErrorCode);
     HelpFunctions_DllExport int  VCS_GetPortNameSelection(char* DeviceName, char* ProtocolStackName, char* InterfaceName, int StartOfSelection, char* pPortSel, unsigned short MaxStrSize, int* pEndOfSelection, unsigned int* pErrorCode);
+	HelpFunctions_DllExport int  VCS_ResetPortNameSelection(char* DeviceName, char* ProtocolStackName, char* InterfaceName, DWORD* pErrorCode);
```
Comment (benmaidel) [inline epos_library/include/epos_library/Definitions.h:69]: Tab vs. Spaces

File: epos_library/include/epos_library/Definitions.h [hunk 11]
```diff
@@ -99,10 +98,12 @@
     Configuration_DllExport int  VCS_SetIncEncoderParameter(void* KeyHandle, unsigned short NodeId, unsigned int EncoderResolution, int InvertedPolarity, unsigned int* pErrorCode);
     Configuration_DllExport int  VCS_SetHallSensorParameter(void* KeyHandle, unsigned short NodeId, int InvertedPolarity, unsigned int* pErrorCode);
     Configuration_DllExport int  VCS_SetSsiAbsEncoderParameter(void* KeyHandle, unsigned short NodeId, unsigned short DataRate, unsigned short NbOfMultiTurnDataBits, unsigned short NbOfSingleTurnDataBits, int InvertedPolarity, unsigned int* pErrorCode);
+	Configuration_DllExport int  VCS_SetSsiAbsEncoderParameterEx(void* KeyHandle, unsigned short NodeId, unsigned short DataRate, unsigned short NbOfMultiTurnDataBits, unsigned short NbOfSingleTurnDataBits, unsigned short NbOfSpecialDataBits, int InvertedPolarity, unsigned short Timeout, unsigned short PowerupTime, unsigned int* pErrorCode);
```
Comment (benmaidel) [inline epos_library/include/epos_library/Definitions.h:101]: Tab vs. Spaces

**Review by benmaidel — CHANGES_REQUESTED** [review]

**Review by benmaidel — CHANGES_REQUESTED** [review]
It would be better to split coding style changes from functional changes. It is hard to find the important changes.

As this code is mainly copied from somewhere, would't it make sense to just write a specialization for the epos class?

**Comment by jabbenseth** [comment]
close here. restyling first, then I add new functionalities later

File: epos_library/include/epos_library/Definitions.h (hunk 9 above)
Comment (jabbenseth) [reply epos_library/include/epos_library/Definitions.h:61]: this is shit. kudos maxon for messing up

**Comment by jabbenseth** [comment]
I'm really in a state of "fuck this" with this repository. I did some changes needed for the hardware (because right now the only CAN interface is :hankey: and only "works" on the nvidia jetson) -> find a library -> use the library. I updated the  `epos_library` part to a version where EPOS4 Hardware runs, ran into trouble with units (tick/s vs rad/s). I think the changes should be documented (like I anticipated with this PR) but are somehow breaking for the original Interface from river-labs, but will be obsolete once CAN works.

Again. I agree with all of the comments/change requests, but will not fix existing code. I messed up (fixed) the messed up formatting for one file. Sorry about that (just review with `?w=1´)

**Comment by ipa-fez** [comment]
@ipa-jba what's your comment about the `udev` stuff?

Resolve that and delete the big chunk of commented out code and it's OK for me...

**Comment by jabbenseth** [comment]
@ipa-fez I'm on it (clamping as well :smile: )
The next pr will restyle the :hankey: out of here

**Review by ipa-fez — APPROVED** [review]

File: epos_library/include/epos_library/Definitions.h (hunk 9 above)
Comment (benmaidel) [reply epos_library/include/epos_library/Definitions.h:61]: ?

Comment (jabbenseth) [reply epos_library/include/epos_library/Definitions.h:61]: the whole Definitions.h file is copied from the official library releasey by maxonmotor. I did not do any modifications -> I don't feel responsible or even see the need to style this --> thanks to maxonmotor for messing up the style and me receiving the blame. 

See https://www.maxonmotor.de/medias/sys_master/root/8825412976670/EPOS-Linux-Library-En.zip

**Comment by benmaidel** [comment]
Man I've never seen so many :hankey: 's in one PR :)
Eat that :hankey: and get your approve

**Review by benmaidel — APPROVED** [review]

**Pull request merged**